File: 0_3.leetcode/java/Remove_Nth_Node_From_End_Of_List.java
public class Remove_Nth_Node_From_End_Of_List {
    static class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode start = new ListNode();
        ListNode slow = start;
        ListNode fast = start;
        start.next = head;

        // go
        for (int i=0;i<=n;i++) {
            if (fast == null) break;
            fast = fast.next;
        }

        while (fast != null) {
            fast = fast.next;
            slow = slow.next;
        }

        //remove
        slow.next = slow.next.next;

        return start.next;
    }
}

File: 0_3.leetcode/kotlin/src/main/kotlin/Remove_Nth_Node_From_End_Of_List.kt
class ListNode(var `val`: Int) {
    var next: ListNode? = null
}

fun removeNthFromEnd(head: ListNode?, n: Int): ListNode? {
    val start = ListNode(0)
    var slow: ListNode? = start
    var fast: ListNode? = start
    start.next = head

    for (i in 0..n) {
        fast = fast?.next
    }

    while (fast != null) {
        fast = fast.next
        slow = slow?.next
    }

    slow?.next = slow?.next?.next

    return start.next
}
